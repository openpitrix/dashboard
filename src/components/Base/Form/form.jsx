import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './index.scss';

export default class Form extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    data: PropTypes.object,
    onSubmit: PropTypes.func
  };

  static defaultProps = {
    className: ''
  };

  constructor(props) {
    super(props);
    this._formData = props.data || {};
  }

  componentWillReceiveProps(nextProps) {
    if ('data' in nextProps) {
      this._formData = nextProps.data || {};
    }
  }

  handleSubmit = e => {
    e.preventDefault();
    // validator

    this.props.onSubmit(this._formData);
  };

  getData() {
    return this._formData;
  }

  render() {
    const { className, children } = this.props;

    const classNames = classnames(styles.form, className);

    const childNodes = React.Children.map(children, child => React.cloneElement(child, {
      ...child.props,
      formData: this._formData
    }));

    return (
      <form className={classNames} onSubmit={this.handleSubmit}>
        {childNodes}
      </form>
    );
  }
}
