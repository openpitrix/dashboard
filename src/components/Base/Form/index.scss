@import '~scss/variables';

.item {
  margin-bottom: $base-margin + 18;
  .label{
    width: 150px;
    line-height: 16px;
    font-weight: bold;
    text-align: right;
    color: $title-color;
  }
  .wrapper {
    display: inline-block;
    position: relative;
    min-width: 276px;
  }
  .help {
    position: absolute;
    left: 0;
    bottom: -18px;
    font-size: 12px;
    color: $second-text-color;
  }
}

.section {
  margin: $base-margin;
}

.sectionTitle {
  margin-bottom: 32px;
  line-height: 19px;
  font-size: 16px;
  font-weight: bold;
  color: $title-color;
}