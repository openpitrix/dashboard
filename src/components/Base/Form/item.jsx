import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './index.scss';

export default class FormItem extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    formData: PropTypes.object,
    help: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    label: PropTypes.string,
    labelClass: PropTypes.string,
    noLabel: PropTypes.bool,
    required: PropTypes.bool,
    wrapperClass: PropTypes.string
  };

  static defaultProps = {
    className: '',
    noLabel: false
  };

  handleValueChange = e => {
    const { formData, children } = this.props;
    const { name, onChange } = children.props;

    // validator
    if (name) {
      formData[name] = e.target.value;
    }

    onChange && onChange(e);
  };

  render() {
    const {
      className,
      children,
      label,
      help,
      formData,
      labelClass,
      wrapperClass,
      noLabel
    } = this.props;
    const { name } = children.props;

    const classNames = classnames(styles.item, className);
    const labelClassNames = classnames(styles.label, labelClass);
    const wrapperClassNames = classnames(styles.wrapper, wrapperClass);

    const childNode = React.cloneElement(children, {
      ...children.props,
      onChange: this.handleValueChange,
      defaultValue: formData[name],
      id: name
    });

    return (
      <div className={classNames}>
        {!noLabel && (
          <label htmlFor={name} className={labelClassNames}>
            {label}
          </label>
        )}
        <div className={wrapperClassNames}>
          {childNode}
          {help && <span className={styles.help}>{help}</span>}
        </div>
      </div>
    );
  }
}
