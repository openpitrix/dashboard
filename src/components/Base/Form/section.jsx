import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './index.scss';

export default class FormItem extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.arrayOf(PropTypes.node)
    ]),
    className: PropTypes.string,
    formData: PropTypes.object,
    title: PropTypes.string
  };

  static defaultProps = {
    className: ''
  };

  render() {
    const {
      className, children, title, formData
    } = this.props;

    const classNames = classnames(styles.section, className);

    const childNodes = React.Children.map(children, child => React.cloneElement(child, {
      ...child.props,
      formData
    }));

    return (
      <div className={classNames}>
        <p className={styles.sectionTitle}>{title}</p>
        {childNodes}
      </div>
    );
  }
}
