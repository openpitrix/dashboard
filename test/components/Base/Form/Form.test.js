import Form from 'components/Base/Form/form';

describe('Base/Form', () => {
  it('basic render', () => {
    const wrapper = render(<Form data={{ name: 'test' }}/>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
