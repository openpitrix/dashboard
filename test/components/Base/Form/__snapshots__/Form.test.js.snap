// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Form basic render 1`] = `
<form
  class="form"
/>
`;
